src/lib/util/options.cpp

```diff
@@ -64,244 +64,484 @@ namespace
 //-------------------------------------------------
 
-core_options::entry::entry(const char *name, const char *description, uint32_t flags, const char *defvalue)
-	: m_next(nullptr),
-		m_flags(flags),
-		m_error_reported(false),
-		m_priority(OPTION_PRIORITY_DEFAULT),
-		m_description(description),
-		m_changed(false)
+core_options::entry::entry(std::vector<std::string> &&names, core_options::option_type type, const char *description)
+	: m_names(std::move(names))
+	, m_priority(OPTION_PRIORITY_DEFAULT)
+	, m_type(type)
+	, m_description(description)
 {
-	// copy in the name(s) as appropriate
-	if (name != nullptr)
+	if (type == option_type::HEADER)
+		assert(m_names.size() == 0);
+	else
+		assert(m_names.size() > 0);
+}
+
+core_options::entry::entry(std::string &&name, core_options::option_type type, const char *description)
+	: entry(std::vector<std::string>({ std::move(name) }), type, description)
+{
+}
+
+
+//-------------------------------------------------
+//  entry - destructor
+//-------------------------------------------------
+
+core_options::entry::~entry()
+{
+}
+
+
+//-------------------------------------------------
+//  entry::value
+//-------------------------------------------------
+
+const char *core_options::entry::value() const
+{
+	// returning 'nullptr' from here signifies a value entry that is essentially "write only"
+	// and cannot be meaningfully persisted (e.g. - a command or the software name)
+	return nullptr;
+}
+
+
+//-------------------------------------------------
+//  entry::set_value
+//-------------------------------------------------
+
+void core_options::entry::set_value(std::string &&newvalue, int priority_value, bool always_override)
+{
+	// it is invalid to set the value on a header
+	assert(type() != option_type::HEADER);
+
+	// only set the value if we have priority
+	if (always_override || priority_value >= priority())
 	{
-		// first extract any range
-		std::string namestr(name);
-		int lparen = namestr.find_first_of('(',0);
-		int dash = namestr.find_first_of('-',lparen + 1);
-		int rparen = namestr.find_first_of(')',dash + 1);
-		if (lparen != -1 && dash != -1 && rparen != -1)
+		internal_set_value(std::move(newvalue));
+		m_priority = priority_value;
+
+		// invoke the value changed handler, if appropriate
+		if (m_value_changed_handler)
+			m_value_changed_handler(value());
+	}
+}
+
+
+//-------------------------------------------------
+//  entry::set_default_value
+//-------------------------------------------------
+
+void core_options::entry::set_default_value(std::string &&newvalue)
+{
+	// set_default_value() is not necessarily supported for all entry types
+	throw false;
+}
+
+
+//-------------------------------------------------
+//  entry::validate
+//-------------------------------------------------
+
+void core_options::entry::validate(const std::string &data)
+{
+	float fval;
+	int ival;
+	int minimum_integer, maximum_integer;
+	float minimum_float, maximum_float;
```

Comment:
Scope some of these down?  I'm pretty sure at least some of them are only required for some types.